### PR TITLE
feat: add queue action

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/queue.py [hunk 1]
```diff
@@ -118,16 +169,39 @@ def add_pull(self, ctxt: context.Context, config: QueueConfig) -> None:
         )
 
         if added:
+            if self.train:
+                # NOTE(sileht): We don't care about checking the queue name here, but we
+                # please mypy
```
Comment (jd) [inline mergify_engine/queue.py:174]: maybe go with:
```suggestion
            if not self.train:
                # NOTE(sileht): We don't care about checking the queue name here, but we
                # please mypy
                raise RuntimeError(…)
```
rather than failing silently?
If we ever modify this code later we might have a bug and miss it.

File: mergify_engine/merge_train.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        )
+
+    def _save(self):
+        prepared = self.Serialized(
+            waiting_pulls=self._waiting_pulls,
+            current_base_sha=self._current_base_sha,
+            cars=[c.serialized() for c in self._cars],
+        )
+        raw = json.dumps(prepared)
+        self.redis.set(self._get_train_queue_key(), raw)
```
Comment (jd) [inline mergify_engine/merge_train.py:337]: This never expires, when is it deleted?

File: mergify_engine/merge_train.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+
+    def get_car_by_tmp_pull(self, ctxt: context.Context) -> typing.Optional[TrainCar]:
+        for car in self._cars:
+            if car.tmp_pull_request_number == ctxt.pull["number"]:
+                return car
+        return None
```
Comment (jd) [inline mergify_engine/merge_train.py:343]: ok but now you can do:
```suggestion
return first.first(car for car in self._cars if car.tmp_pull_request_number == ctxt.pull["number"])
```
😉

File: mergify_engine/merge_train.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+    def refresh(self) -> None:
+        self._populate_cars()
+        self._save()
+        self.log.info("refreshed train cars")
+
+    def reset(self) -> None:
+        self._slice_cars_at(0)
+        self._populate_cars()
+        self._save()
+        self.log.info("resetted train cars")
```
Comment (jd) [inline mergify_engine/merge_train.py:354]:
```suggestion
        self.log.info("train cars reset")
```

File: mergify_engine/merge_train.py [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+    def get_car_by_tmp_pull(self, ctxt: context.Context) -> typing.Optional[TrainCar]:
+        for car in self._cars:
+            if car.tmp_pull_request_number == ctxt.pull["number"]:
+                return car
+        return None
+
+    def refresh(self) -> None:
+        self._populate_cars()
+        self._save()
+        self.log.info("refreshed train cars")
```
Comment (jd) [inline mergify_engine/merge_train.py:348]:
```suggestion
        self.log.info("train cars refreshed")
```

File: mergify_engine/merge_train.py [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+        pull: github_types.GitHubPullRequest = self.train.client.item(
+            f"/repos/{self.train.owner_login}/{self.train.repo_name}/pulls/{self.pull_request_number}"
+        )
+
+        sub = asyncio.run(
+            subscription.Subscription.get_subscription(self.train.owner_id)
+        )
+        original_ctxt = context.Context(self.train.client, pull, sub, [])
+
+        title = "This pull request cannot tested to be merged"
```
Comment (jd) [inline mergify_engine/merge_train.py:186]:
```suggestion
        title = "This pull request cannot be queued for merge"
```

File: mergify_engine/merge_train.py [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+        summary = f"This pull request can't be tested to be merged because: {exception.message}"
```
Comment (jd) [inline mergify_engine/merge_train.py:187]:
```suggestion
        summary = f"This pull request can't be queued for merge: {exception.message}"
```

File: mergify_engine/merge_train.py [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+        tmp_pull_ctxt: context.Context,
+        queue_rule: typing.Optional[rules.EvaluatedQueueRule],
+        status: check_api.Conclusion,
+        deleted: bool,
+    ) -> None:
+        if status == check_api.Conclusion.SUCCESS:
+            title = f"The pull request #{self.pull_request_number} can be merged"
+        elif status == check_api.Conclusion.PENDING:
+            title = (
+                f"The pull request #{self.pull_request_number} is tested to be merged"
```
Comment (jd) [inline mergify_engine/merge_train.py:217]:
```suggestion
                f"The pull request #{self.pull_request_number} is mergeable"
```

File: mergify_engine/merge_train.py [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+        )
+
+        sub = asyncio.run(
+            subscription.Subscription.get_subscription(self.train.owner_id)
+        )
+        original_ctxt = context.Context(self.train.client, pull, sub, [])
+        title = f"The pull request is tested against {original_ctxt.pull['base']['ref']} to be merged"
+        if deleted:
+            # TODO(sileht): display train cars ?
+            summary = "The pull request is going to be tested soon"
```
Comment (jd) [inline mergify_engine/merge_train.py:260]:
```suggestion
            summary = "The pull request is going to be queued soon"
```

File: mergify_engine/merge_train.py [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+        else:
+            summary = f"You can follow the pull request testing here: {tmp_pull_ctxt.pull['html_url']}"
```
Comment (jd) [inline mergify_engine/merge_train.py:262]:
```suggestion
            summary = f"You can follow the pull request queue here: {tmp_pull_ctxt.pull['html_url']}"
```

File: mergify_engine/merge_train.py [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+                title=title,
+                summary=summary,
+            ),
+        )
+
+        asyncio.run(github_events.send_refresh(original_ctxt.pull))
+
+    def update_summaries(
+        self,
+        tmp_pull_ctxt: context.Context,
```
Comment (jd) [inline mergify_engine/merge_train.py:208]: Not a fan of the `tmp` here. Can we find a better name?

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/queue.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/queue.py:174]: The comment is wrong, it's a leftover of another `if` and I changed the code upstream (L156)  to avoid the `None` checking everywhere to please just mypy.

File: mergify_engine/merge_train.py [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+
+
+@dataclasses.dataclass
+class Train:
+    client: github.GithubInstallationClient
+    redis: redis.Redis
+    owner_id: int
+    owner_login: str
+    repo_id: int
+    repo_name: str
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/merge_train.py:291]: No, but we can use IdType and Name variant now :)

File: mergify_engine/merge_train.py [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+
+    @classmethod
+    def deserialized(cls, train: "Train", data: "TrainCar.Serialized") -> "TrainCar":
+        return cls(train, **data)
+
+    def _get_branch_name(self, *, escaped=False):
+        if escaped:
+            ref = parse.quote(self.train.ref, safe="")
+        else:
+            ref = self.train.ref
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/merge_train.py:85]: I removed this method.

File: mergify_engine/merge_train.py [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+    def log(self):
+        return daiquiri.getLogger(
+            __name__,
+            gh_owner=self.owner_login,
+            gh_repo=self.repo_name,
+            gh_branch=self.ref,
+            max_size=self.max_size,
+            train_cars=[c.pull_request_number for c in self._cars],
+            train_waiting_pulls=[wp.pull_request_number for wp in self._waiting_pulls],
+        )
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/merge_train.py:328]: If I do that `train_cars` and `train_waiting_pulls` will be wrong.

File: mergify_engine/merge_train.py (hunk 2 above)
Comment (sileht) [reply mergify_engine/merge_train.py:337]: Fixed

File: mergify_engine/merge_train.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/merge_train.py:343]: This method should go in stdlib !

File: mergify_engine/merge_train.py (hunk 9 above)
Comment (sileht) [reply mergify_engine/merge_train.py:260]: Rewording:

queued can be confusing as the PR is first queued in the Queue and then queued in the Train.

File: mergify_engine/merge_train.py (hunk 11 above)
Comment (sileht) [reply mergify_engine/merge_train.py:208]: I rename some attributes, let me  known what you think.

File: mergify_engine/merge_train.py (hunk 9 above)
Comment (jd) [reply mergify_engine/merge_train.py]: Understood. In that case `queue` is not the right word indeed.
What about using `embarked`?
We could use `embark` and `disembark` as the verbs to described what happens to PRs that are in queue and now art part of a train.

**Review by jd — DISMISSED** [review]
see wording comment :)

File: mergify_engine/merge_train.py (hunk 9 above)
Comment (sileht) [reply mergify_engine/merge_train.py]: far better :smile:

File: mergify_engine/merge_train.py [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+
+        try:
+            pulls_for_title = ", ".join(
+                [
+                    f"#{p}"
+                    for p in self.parent_pull_request_numbers
+                    + [self.user_pull_request_number]
+                ]
+            )
+            title = f"merge-queue: testing {self.initial_current_base_sha[:7]}, {pulls_for_title} together"
```
Comment (jd) [inline mergify_engine/merge_train.py:139]: should we use "embarking" here?

File: mergify_engine/merge_train.py [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+
+        pulls_for_title = ", ".join(
+            [
+                f"#{p}"
+                for p in self.parent_pull_request_numbers
+                + [self.user_pull_request_number]
+            ]
+        )
+        summary = (
+            f"Testing {self.initial_current_base_sha[:7]}, {pulls_for_title} together"
```
Comment (jd) [inline mergify_engine/merge_train.py:234]: Embarking?

File: mergify_engine/tests/conftest.py [hunk 17]
```diff
@@ -26,3 +27,17 @@ def logger_checker(request, caplog):
 def setup_new_event_loop():
     # ensure each tests have a fresh event loop
     asyncio.set_event_loop(asyncio.new_event_loop())
+
+
+@pytest.fixture(autouse=True)
+async def setup_aredis():
```
Comment (jd) [inline mergify_engine/tests/conftest.py:33]:
```suggestion
async def setup_aredis_for_cache():
```

**Review by jd — CHANGES_REQUESTED** [review]
wording

File: docs/source/queues.rst [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+allows the pull request to enter in the queue when all conditions match. If
+the ``queue`` action conditions do not match anymore the pull request is
+removed from the merge queue. When a pull request is the next one to be
+merged, Mergify looks at the seconds set of conditions defined this time in
+``queue_rules`` and wait that they all match to merge the pull request.
+
+When multiple queues are defined, they are processed one after the other in the
+order they are defined in the configuration. When the first pull request in the
+first queue is waiting to be merged, the remaining queues are blocked.
+
```
Comment (jd) [inline docs/source/queues.rst:69]: we should add something about the cases where the PR is removed from the queue: either by being merged or by not matching the conditions anymore.

File: mergify_engine/merge_train.py [hunk 19]
```diff
@@ -0,0 +1,522 @@
...
+        pull: github_types.GitHubPullRequest = self.train.client.item(
+            f"/repos/{self.train.owner_login}/{self.train.repo_name}/pulls/{self.user_pull_request_number}"
+        )
+
+        sub = await subscription.Subscription.get_subscription(self.train.owner_id)
+        original_ctxt = context.Context(self.train.client, pull, sub, [])
+        title = f"The pull request is embarked with {pulls_for_title} to be merged"
+        if will_be_reset:
+            # TODO(sileht): display train cars ?
+            summary = "The pull request is going to be re-embarked soon:  {tmp_pull_ctxt.pull['html_url']}"
```
Comment (jd) [inline mergify_engine/merge_train.py:262]:
```suggestion
            summary = "The pull request is going to be re-embarked soon: {tmp_pull_ctxt.pull['html_url']}"
```

**Review by jd — CHANGES_REQUESTED** [review]

File: docs/source/queues.rst [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+Working with Queue Rules
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+With a ``queue`` action, a pull request has to match two sets of conditions
+to get merged. The first set of conditions described in ``pull_request_rules``
+allows the pull request to enter in the queue when all conditions match. If
+the ``queue`` action conditions do not match anymore the pull request is
+removed from the merge queue. When a pull request is the next one to be
+merged, Mergify looks at the seconds set of conditions defined this time in
```
Comment (jd) [inline docs/source/queues.rst:74]:
```suggestion
removed from the merge queue. When a pull request is at the top of the queue, Mergify looks at the seconds set of conditions defined this time in
```

File: mergify_engine/merge_train.py [hunk 21]
```diff
@@ -0,0 +1,505 @@
...
+        *,
+        queue_rule: typing.Optional[rules.EvaluatedQueueRule] = None,
+        will_be_reset: bool = False,
+    ) -> None:
+        if status == check_api.Conclusion.SUCCESS:
+            title = f"The pull request #{self.user_pull_request_number} is mergeable"
+        elif status == check_api.Conclusion.PENDING:
+            title = f"The pull request #{self.user_pull_request_number} is embarked for merge"
+        else:
+            title = f"The pull request #{self.user_pull_request_number} is disembark as it cannot be merged"
```
Comment (jd) [inline mergify_engine/merge_train.py:224]:
```suggestion
            title = f"The pull request #{self.user_pull_request_number} cannot be merged and has been disembarked"
```

File: mergify_engine/merge_train.py [hunk 22]
```diff
@@ -0,0 +1,505 @@
...
+        try:
+            await self.train.repository.installation.client.delete(
+                f"/repos/{self.train.repository.installation.owner_login}/{self.train.repository.name}/git/refs/heads/{branch}"
+            )
+        except http.HTTPNotFound:
+            pass
+
+    async def _report_failure(self, exception: http.HTTPClientSideError) -> None:
+        title = "This pull request cannot be embarked for merge"
+        summary = f"This pull request can't be embarked for merge: {exception.message}"
```
Comment (jd) [inline mergify_engine/merge_train.py:181]: repetitive?
```suggestion
        title = "This pull request cannot be embarked for merge"
        summary = exception.message
```
?

File: mergify_engine/tests/functional/base.py [hunk 23]
```diff
@@ -444,6 +444,7 @@ def github_aclient(owner=None, auth=None):
         )
 
         self.cli_integration = github.aget_client(config.TESTING_ORGANIZATION)
+        self.addAsyncCleanup(self.cli_integration.aclose)
```
Comment (jd) [inline mergify_engine/tests/functional/base.py:447]: unrelated?

**Review by jd — CHANGES_REQUESTED** [review]
just a couple of type otherwise good to go

File: docs/source/queues.rst [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+
+In that case, Mergify creates a new pull request ``Z`` with the content of
+``12345678`` + ``C`` + ``D``.
+
+In the case where a new commit is pushed on the base branch, Mergify resets the
+merge train and starts over the process.
+
+If a pull request embarked doesn't match the ``queue_rules`` anymore, this one
+is retired from the merge train and all pull requests embarked after this
+one are dis-embarked and re-embarked.
```
Comment (jd) [inline docs/source/queues.rst:64]:
```suggestion
If an embarked pull request doesn't match the ``queue_rules`` anymore, it
is removed from the merge train. All pull requests embarked after it
are disembarked and re-embarked.
```

**Review by jd — DISMISSED** [review]

File: docs/source/queues.rst (hunk 24 above)
Comment (jd) [reply docs/source/queues.rst:64]: Can you just update this ^ ?

**Review by jd — DISMISSED** [review]
Just a doc thing

**Review by jd — APPROVED** [review]
